README.md
# Mr-Robot backup python script - Dump database, compact, incremental and copy to remote server.
# Dedicated the serie Mr Robot

# download from github
$ git clone https://github.com/teagom/mr-robot.git

$ cd mr-robot

# to copy a example.py
$ cp example.py config_app_name.py

$ chmod 700 config_app_name.py

$ vim config_app_name.py

# new setting
$ cp setting.py.DIST settings.py

$ chmod 700 setting.py

$ vim setting.py

# Security content of this script because user and password.
# Keep safe the content of scripts.
$ chmod 700 mr-robot.py

# run mr-robot backup script
$ mr-robot.py config_app_name.py

--- The ideia of this script is make a full backup for server or app.
	
	Step by step to facility the backup life!
	1	Dump database and compact
	2 	Compact file and folder
	3	Increment file and folder
    4       Copy backup to remote server
    5	Set permission for backup folder
    6	Send a mail after finish backup

	--- struct of file and folder:

        		backup-folder/
                            app_config_name/
                                            incremental/
                                            date-time/ ( maybe more than one backup per day )
                                                        db.ext
                                                        files.ext
                                                        log
                                                        log.err
	
--- config array

	--- base
	0	config or app name ( string )
	1	frequency of backup (once|daily|week-full|month-full)

        frequency
            date-time folder will be created based in frenquency choice.
            folder name will be created based in date-time
            
		crontab will be define backup time. example: two times per day, 12:01 and 00:01 
		        
		once: backup of 1 day ago
                will overwrite last backup and incremental new files, folders and altered files.
                output format:
                            /backup/config_name/00h01m
                            /backup/config_name/12h01m

            daily: never repeat, never overwrite
                    incremental new files, folders and altered files
                    output format:
                            /backup/config_name/31_12_2015-00h01m
                            /backup/config_name/31_12_2015-12h01m
                            /backup/config_name/01_01_2016-00h01m
                            /backup/config_name/01_01_2016-12h01m
                            /backup/config_name/...-00h01m
                            /backup/config_name/...-12h01m

            week-full: sunday to saturday / backup of 7 days ago
                incremental new files, folders and altered files
                output format:
                                /backup/config_name/sunday_00h01m
                                /backup/config_name/sunday_12h01m
                                /backup/config_name/..._00h01m
                                /backup/config_name/..._12h01m
                                /backup/config_name/saturday_00h01m
                                /backup/config_name/saturday_12h01m
            
		month-full: 01 to 31
			output format:
                            /backup/config_name/01_00h01m
                            /backup/config_name/01_12h01m
                            /backup/config_name/..._00h01m
                            /backup/config_name/..._12h01m
                            /backup/config_name/31_00h01m
                            /backup/config_name/31_12h01m

    2   (True|False) Delete local or temporary backup after copy to remote server
            If log.err are not empty will be not dropped and a will be send a email to report error.
            Is possible to make backup two times, in the external HD and remote server, use False.

    3   (string) full path to backup folder
    4   (True|False) sendmail after finish backup
    5   (string) administrator email address
    
    --- compress settings
    6   (string) extension file of compactor program
	7   (string) full path to binary and parameters to compact
    8   (string) full path to binary and parameters to test
    
    --- database
    10  (True|False) Dump database
    11  (string) (postgres|mysql) database server
    12  (string) hostname or ip address
    13  (string) database name
    14  (string) user
    15  (string) password
        
    --- compress file and folder
    20  (True|False) compress file and folder
    21  (array)	Include theses paths to folder or file.
            ['/etc/apache','/var/www/blog']
    22  (array) exclude ['/var/www/blog/temp/*']
            this folder is not be included
            use * to exclude all file and folder

    --- incremental file and folder
    30  (True|False) incremental file and folder
	31  (string) incremental folder name
    32  (array) [ '/etc/' , ' /var/log' , ' /backup/ ' ]
        include file and folder
	33  exclude	[ 'cacerts','bind9' ]
            theses files and folders is not be included
            Declared just file name or folder name without path to.

    --- copy to remote server
    40  (True|False) to copy to remote server
    41  (string) ip or hostname
    42  (string) port port number
    43  (string) backup folder name in the remote server
    44  (string) rsync parameters '-arvz --progress --partial'
    45  (string) (password|pemfile|authorized)
    46  (string) full path to ssh pem file if [45] is pemfile
    
    # sshpass are required - $ sudo apt-get install sshpass
    # password - First connection have to be confirmed written "yes" in a line command.
    # backup@host$ ssh -p 222 backup@192.168.0.254
    # The authenticity of host '([192.168.0.254]:222)' can't be established.
    # ECDSA key fingerprint is bb:ee:cc:ee:aa:ff:gg.
    # Are you sure you want to continue connecting (yes/no)?
    47  (string) username
    48  (string) password

    --- permission of backup
    50  (string) owner:group chown
    51  (string) 700         chmod


changelog.txt
--- 27 March 2017 - Tiago Souza Moraes
	added
		set permission of backup files after finish, chmod and chown.
			separete permission for db, compact files and incremental files.
		resume of backup, total size, amount of files, total time to make backup.
			total transfer to remote server


external.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import os
from datetime import datetime
from setting import debug, sender, smtpserver, username, password

def cmd_run(cmd, log=False, log_err=False, action=False):

    if debug:
        print
        print 'You are seeing this message because debug is True'
        print cmd

    if action:
        os.system("echo '--- %s' 1>>%s 2>>%s" % (action, log, log_err) )

    if not log and not log_err:
        os.system(cmd)
    else:
        os.system(cmd + "> /dev/null 1>>%s 2>>%s" % (log, log_err) )


def sendmail(to, app, attach0=False, attach1=False):

    from smtplib import SMTP_SSL as SMTP
    from email.MIMEText import MIMEText

    destination = [to]

    # read attach
    contentattach = ''
    if attach0 and not os.stat("%s" % attach0).st_size == 0: 
        fp = open(attach0,'rb')
        contentattach += '------------------ Error begin\n\n'
        contentattach += fp.read()
        contentattach += '\n------------------ Error end'
        fp.close()

    if attach1 and not os.stat("%s" % attach1).st_size == 0: 
        fp = open(attach1,'rb')
        contentattach += '\n\n------------------ Success begin\n\n'
        contentattach += fp.read()
        contentattach += '\n------------------ Success end'
        fp.close()

    msg = MIMEText(contentattach, 'plain')
    msg['Subject'] = "Mr.Script %s" % app
    msg['From'] = sender
                    
    try:
        conn = SMTP(smtpserver)
        conn.set_debuglevel(False)
        conn.login(username, password)
        conn.sendmail(sender, destination, msg.as_string())
        conn.close()
    except:
        print '*** Error trying send a mail. Check settings.'


# date format
def dateformat(o):
    '''
        return date format based in frequency set
    '''
    if o == 'daily':
        return datetime.now().strftime("%A-%d_%m_%Y-%HH%MM").lower() # sunday-28_07_2014-14h00m (weekday_day of month-month-year_hour-min)

    if o == 'month-full':
        return datetime.now().strftime("%d-%HH%MM") # 28-14h00m (day of month_hour_min)

    if o == 'week-full':
        return datetime.now().strftime("%A-%HH%MM").lower() # sunday-14h00m (weekday_hour-min)

    if o == 'once':
        return datetime.now().strftime("%HH%MM") # 14h00m (hour-min)


mr-robot.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import argparse
from external import cmd_run, sendmail, dateformat

'''
    Mr-Robot-Backup. Dedicated the serie Mr Robot.

    author:
        Tiago de Souza Moraes
        teagom@gmail.com

    required lib:
        argparse
        sshpass

    1 - Dump and compact Data Base
    2 - Compact file and folder
    3 - Sync increment file and folder to incremental folder
    4 - copy backup to remote server
    5 - Set permission of folder and file

    struct of backup folder, see array position, config[x]
                    /x[3]/ (root backup folder)

                        x[0]/ (app name folder)

                            data-time/
                                    increment/
                                    log
                                    log.err
                                    db.ext
                                    files.ext
'''

parser = argparse.ArgumentParser()
parser.add_argument('integers', metavar='file.py', type=str, nargs='+', help='Python file, parameters to make backup. See example.py')
args = parser.parse_args()

# main code 
for cfg in args.integers:

    # import
    module = __import__(cfg.replace('.py', ''))
    x = module.config

    print
    print '# # # # # # # # # # # # # # # # # # # # # # # # #'
    print '# Mr Robot Backup. Dump, compact and increment  #'
    print '# Running %s' % cfg
    print '# # # # # # # # # # # # # # # # # # # # # # # # #'

    # # # # # # # # # # # # # # # # # # # # # # # #
    # create destiny folder and logs

    # root folder to app backup
    base_backup_app = '%s/%s' % (x[3], x[0])

    # once backup
    if x[1] == 'once':
        cmd_destiny_backup_app = 'mkdir -p %s/%s' % (base_backup_app, dateformat('once'))
        destiny_backup_app = '%s/%s' % (base_backup_app, dateformat('once'))

    # daily backup
    if x[1] == 'daily':
        cmd_destiny_backup_app = 'mkdir -p %s/%s' % (base_backup_app, dateformat('daily'))
        destiny_backup_app = '%s/%s' % (base_backup_app, dateformat('daily'))

    # weekly backup
    if x[1] == 'week-full':
        cmd_destiny_backup_app = 'mkdir -p %s/%s' % (base_backup_app, dateformat('week-full'))
        destiny_backup_app = '%s/%s' % (base_backup_app, dateformat('week-full'))

    # monthly backup
    if x[1] == 'month-full':
        cmd_destiny_backup_app = 'mkdir -p %s/%s' % (base_backup_app, dateformat('month-full'))
        destiny_backup_app = '%s/%s' % (base_backup_app, dateformat('month-full'))

    # call shell command
    cmd_run(cmd_destiny_backup_app)

    # create log files
    log = "%s/log" % destiny_backup_app
    log_err = "%s/log.err" % destiny_backup_app

    clean_log = 'rm -f %s' % log
    clean_log_err = 'rm -f %s' % log_err

    create_log = 'touch %s' % log
    create_log_err = 'touch %s' % log_err

    cmd_run(clean_log)
    cmd_run(clean_log_err)

    cmd_run(create_log)
    cmd_run(create_log_err)


    # # # # # # # # # # # # # # # # # # # # # # # #
    # dump data base and compress
    if x[10]:

        # DB backup file name 
        outputfile_db_tmp = u'%s/db.sql' % ( destiny_backup_app )
        outputfile_db = u'%s/db.%s' % ( destiny_backup_app, x[6] )

        # postgres
        if x[11] == 'postgres' :
            dump = 'export PGPASSWORD=%s\npg_dump --username=%s -h %s %s > %s' % ( x[15], x[14], x[12], x[13], outputfile_db_tmp )

        # mysql
        if x[11] == 'mysql' :
            dump = 'mysqldump --password="%s" -h %s -u %s %s >  %s' % ( x[15], x[12], x[14], x[13], outputfile_db_tmp )

        if x[11]:
            print '> Dump and compact a data base                 '
            compact = '%s %s %s' % ( x[7], outputfile_db, outputfile_db_tmp )
            cmd_run(dump)
            cmd_run(compact, log, log_err, 'dump-database-compact')

            if x[8]:
                test = '%s %s' % ( x[8], outputfile_db )
                cmd_run(test, log, log_err,'dump-database-test-zip')

            cmd_run('rm -f %s' % (outputfile_db_tmp), log, log_err, 'dump-database-delete-sql')

    # # # # # # # # # # # # # # # # # # # # # # # #
    # compress file and folder

    # output file
    outputfile_backup = u'%s/files.%s' % ( destiny_backup_app, x[6] )

    if x[20]:
        print '> Compress file and folder'

        # exclude list
        exclude = ''
        for xx in x[22]:
            exclude += "--exclude \"%s\" " % xx

        # include list
        include = ''
        for xx in x[21]:
            include += "%s " % xx

        # compactor line command
        compact = "%s %s %s %s" % ( x[7], outputfile_backup, include, exclude )
        cmd_run(compact, log, log_err, 'file-folder-compact')

        if x[8]:
            test = '%s %s' % ( x[8], outputfile_backup )
            cmd_run(test, log, log_err, 'file-folder-compact-test')

    # # # # # # # # # # # # # # # # # # # # # # # #
    # increment file and folder
    if x[30]:
        print '> Increment file and folder'

        cmd_destiny_backup_app = 'mkdir -p %s/%s' % ( base_backup_app, x[31] )
        cmd_run(cmd_destiny_backup_app, log, log_err, 'incremental-file-folder')

        include = ""
        for xx in x[32]:
            include += "%s " % xx

        exclude = ""
        for xx in x[33]:
            exclude += "--exclude=\'%s\' " % xx

        rsync = "rsync %s %s %s %s/%s" % ( x[44] , exclude , include , base_backup_app , x[31] )
        cmd_run(rsync, log, log_err, 'file-folder-rsync')


    # # # # # # # # # # # # # # # # # # # # # # # #
    # set permission
    # change permission before transfer to preserve
    # permission at remove server.
    if x[50]:
        c = "chown %s %s -R" % (x[50], base_backup_app)
        cmd_run(c, log, log_err, 'permission-chown')

    if x[51]:
        c = "chmod %s %s -R" % (x[51], base_backup_app)
        cmd_run(c, log, log_err, 'permission-chmod')


    # # # # # # # # # # # # # # # # # # # # # # # #
    # transfer all backup to remote server. 
    # rsync from localhost to remote 
    if x[40]:
        print '> Copy backup to remote server'

        if x[45] == "password":
            rsync = "sshpass -p \"%s\" rsync %s -e \"ssh -p %s\" %s %s@%s:%s" % ( x[48] , x[44] , x[42] , base_backup_app , x[47] , x[41] , x[43] )

        if x[45] == "pemfile":
            rsync = "rsync %s -e \"ssh -p %s -i %s\" %s %s@%s:%s" % ( x[44] , x[42] , x[46] , base_backup_app , x[47] , x[41] , x[43] )

        if x[45] == "authorized":
            rsync = "rsync %s -e \"ssh -p %s \" %s %s@%s:%s" % ( x[44] , x[42] , base_backup_app , x[47] , x[41] , x[43] )

        cmd_run(rsync, log, log_err, 'rsync-backup')


    # sendmail
    # if log.err not empty then sendmail to report
    import os

    if x[4] == True or not os.stat("%s" % log_err).st_size == 0: 
        print '> Send a mail'
        sendmail( x[5], x[0], log_err, log )

    if x[2] == True:
        print '> Delete temporary folder or localhost backup'
        clean = 'rm -rf %s' % ( base_backup_app )
        cmd_run(clean, log, log_err, 'delete-temporary-backup-folder')


example.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

'''
    output of struct of files and folders for these settings, see array position, config[x]

                    config[3] - (root backup folder)
                    /media/ext-hd/
                                    configx[0] - (app name folder)
                                    www_server/

                                                config[31] - incremental folder name
                                                incremental/

                                                config[1] - model store backup files / frequency of backup (once|daily|week-full|month-full)
                                                data-time/
                                                            log
                                                            log.err
                                                            db.ext
                                                            files.ext

'''

from setting import model
import copy

# # # # # # # # # # # # # # # # # # # # # # # #
# set app array
# # # # # # # # # # # # # # # # # # # # # # # #

config = copy.deepcopy(model)

# global variables of backup
config[0] = 'www_server' # name of config or app (string) Will be used to create a folder backup name
config[1] = 'week-full' # frequency of backup (once|daily|week-full|month-full)
config[2] = False # delete local/temporary backup after copy to remote server (True|False)
                  # if log.err are not empty will be not dropped and a will be send a email to report error.
                  # I will make backup two times, localhost at external HD and remote server
config[3] = '/media/ext-hd' # full path to backup folder or temporary folder
config[4] = True # sendmail after finish backup, log and log.err in attach.
config[5] = 'user@mail.com' # admin email address

# options to compact file and dump of data base
config[6] = 'zip' # extension (string)
config[7] = '/usr/bin/zip -9r -P test'
            # full path to binary (string)
            # parameters to compact, password(-P passsword), better, faster, recursive, ...
config[8] = '/usr/bin/unzip -t -P test'
            # full path to binary (string)
            # parameters to test and password(-P passsword)

# data base settings
config[10] =  True # dump of database (True|False)
config[11] = 'postgres'  # db server (postgres|mysql)
config[12] = '127.0.0.1' # host
config[13] = 'portal_db' # db name
config[14] = 'portal_user' # user
config[15] = 'portal_pass' # pass

# backup files
config[20] = True # compact file and folder
config[21] = ['/var/www/www.company.com'] # include
# to exclude, use * in the end of path.
config[22] = ['/var/www/www.company.com/temp/*','/var/www/www.company.com/.git/*','/var/www/www.company.com/upload/*'] 

# incremental file and folder
config[30] = True  # incremental using rsync (True|False)
config[31] = 'incremental' # incremental folder name (string)
config[32] = ['/etc/default','/etc/apache','/var/www/www.company.com/upload'] # include file and folder (string)
config[33] = ['google-chrome','cacerts'] # exclude file and folder (string) # this file are inside of /etc/default. Declared just file name or folder name without path.

# copy backup to remote server
'''
    Incremental folder will be not duplicated, will be incremental. :)
'''
config[40] = False # copy backup to remote host using rsync ssh (True|False)
config[41] = '192.168.0.254' # server ip or hostname (string)
config[42] = '222' # ssh port (string)
config[43] = '~/backup' # destiny folder in the remote server (string)
config[44] = '-arvz --progress --partial'  # rsync parameters (string)
config[45] = 'password' # authentication type (string), (password|pemfile|authorized)

# password - First connection have to be confirmed written "yes" in a command line.
# sshpass are required - $ sudo apt-get install sshpass
# ssh -p 222 backup@192.168.0.254
# The authenticity of host '([192.168.0.254]:222)' can't be established.
# ECDSA key fingerprint is bb:ee:cc:ee:aa:ff:gg.
# Are you sure you want to continue connecting (yes/no)? 

config[46] = '/opt/server-pem-2015.pem' # pem file if [45] is pemfile (string)
config[47] = 'user' # user (string) is required for any [45] option
config[48] = 'password' # password (string)

# permission section
config[50] = 'user:group' # (string) owner:group for root backup folder, full path to and recursive mode 
config[51] = '700' # (string) chmod for root backup folder.
